EVG-9005 Feature custom signature annotation

hashtag-smashtag opened the pull request:
https://help.evergage.com/browse/EVG-9005

hashtag-smashtag committed:
EVG-9005 Custom signature annotation support

hashtag-smashtag committed:
EVG-9005 Custom annotation can also support static method for dynamic value.

 Thinking of enum values as string union, specifically, which could not be accomplished as a compile-time constant expression string.

hashtag-smashtag committed:
EVG-9005 separate custom annotations for clarity

hashtag-smashtag committed:
EVG-9005 many improvements

 * refactor annotations for clarity and conciseness
 * configurably support: abstract/default/static/other methods
 * support sorting methods in addition to properties, stabilizes testing as well
 * null protection around discoveredClasses

hashtag-smashtag committed:
EVG-9005 bump version to -evg2

ispringer commented on typescript-generator-core/src/main/java/cz/habarta/typescript/generator/CustomSignatureTypeProcessor.java:
```diff
@@ -0,0 +1,83 @@
+        Class<?> declaringClass = member.getDeclaringClass();
+        return Arrays.stream(annotations)
+                .map(annotation -> {
+                    if (annotation instanceof TypeScriptSignature) {
```
I'd extract processTypeScriptSignatureAnnotation and processTypeScriptSignatureViaStaticMethodAnnotation methods.

ispringer commented on typescript-generator-core/src/main/java/cz/habarta/typescript/generator/CustomSignatureTypeProcessor.java:
```diff
@@ -0,0 +1,83 @@
+                            clazz = declaringClass;
+                        }
+                        try {
+                            Method method = clazz.getDeclaredMethod(staticMethodName);
```
Maybe a bit more error handling here - if method does not exist, has params, etc.

hashtag-smashtag replied:
I (briefly) thought about it but it's in a try/catch and any such issue would at least be wrapped within a friendly message and provide the cause... I suspect it's good enough until we encounter otherwise

ispringer commented on typescript-generator-core/src/main/java/cz/habarta/typescript/generator/CustomSignatureTypeProcessor.java:
```diff
@@ -0,0 +1,83 @@
+                        return null;
+                    }
+                })
+                .filter(Objects::nonNull)
```
Maybe log an error or throw an exception if there are multiple results, which would mean the type was annotated with both TypeScriptSignature and TypeScriptSignatureViaStaticMethod.

hashtag-smashtag replied:
Good suggestion, will do

hashtag-smashtag committed:
EVG-9005 fix method skip logic

ispringer commented on typescript-generator-core/src/main/java/cz/habarta/typescript/generator/TypeScriptSignature.java:
```diff
@@ -0,0 +1,24 @@
+    /** @return The custom TypeScript signature to output verbatim for this element. */
+    String value();
+
+    /** @return Any classes the generator should processes as if they were 'discovered' in this element. */
```
nice javadoc

ispringer commented on typescript-generator-core/src/main/java/cz/habarta/typescript/generator/TypeScriptSignatureViaStaticMethod.java:
```diff
@@ -0,0 +1,51 @@
+     *         package com.example;
+     *
+     *         public class MyClass {
+     *             @TypeScriptSignatureViaStaticMethod("myEnumCustomSignature")
```
This is nice as it could be used for other use cases, but do you think we should still add a more convenient shortcut annotation?, e.g.:

```
@TypeScriptEnumAsUnion(MyEnum.class) 
public String myEnum;
```

```
@TypeScriptEnumAsUnion 
public MyEnum myEnum;
```

Another option would be to provide a static util method for generating the union that can be reused by TypeScriptSignatureViaStaticMethod impls, e.g.

```
TypeScriptTypeGeneratorUtil#generateUnion(Class<? extends Enum> enumClass)
```

hashtag-smashtag replied:
I've already encountered some of our enums with internal/deprecated values, and a simple enum-class-accepting-annotation wouldn't work for that.  I think we can use our own static method(s), but might as well be within evergage-product and not the generator, imo

ispringer replied:
Ok, maybe this would be a better util method then:

```
TypeScriptTypeGeneratorUtil#generateUnion(EnumSet<? extends Enum> enumValues)
```

Sure, putting the util method in the project code base sounds fine. My main concern is I don't want to repeat the union building code in 20 places.

hashtag-smashtag replied:
yeah, and maybe with an optional param, a function on the enum value, for the case where we call some enum-specific-method on it when name() doesn't suffice

hashtag-smashtag replied:
created the util method(s) over on https://github.com/evergage/evergage-product/pull/507 now on master

hashtag-smashtag committed:
EVG-9005 ImportOrder checkstyle?!  thanks but no thanks

ispringer commented on typescript-generator-core/src/main/java/cz/habarta/typescript/generator/compiler/ModelCompiler.java:
```diff
@@ -869,6 +873,7 @@ private TsModel sortDeclarations(SymbolTable symbolTable, TsModel tsModel) {
         if (settings.sortDeclarations) {
             for (TsBeanModel bean : beans) {
                 Collections.sort(bean.getProperties());
+                Collections.sort(bean.getMethods());
```
Are these lists guaranteed to be mutable?

hashtag-smashtag replied:
Yeah, pretty much ArrayLists, design looks intentionally mutable

hashtag-smashtag committed:
EVG-9005 add null check around emitSAMs

ispringer commented on typescript-generator-core/src/main/java/cz/habarta/typescript/generator/emitter/TsCallableModel.java:
```diff
@@ -54,4 +54,9 @@ public TsType getReturnType() {
+    @Override
+    public int compareTo(TsCallableModel o) {
+        // Maybe also consider parameters/names etc
+        return name.compareTo(o.getName());
```
Maybe use compareToIgnoreCase?

hashtag-smashtag replied:
Since the sort is mostly for stability when testing, I'd rather have it be case sensitive and stable.  For actual use I'd prefer it not sort and spit out declarations in order of appearance.  But internally (at least for 'props') it uses Jackson and it seems to be somewhat random.  Thus the sort for easier testing comparisons

ispringer commented on typescript-generator-core/src/main/java/cz/habarta/typescript/generator/parser/MethodModel.java:
```diff
@@ -9,13 +10,15 @@
     private final List<String> comments;
 
-    public MethodModel(Class<?> originClass, String name, List<MethodParameterModel> parameters, Type returnType, List<String> comments) {
+    public MethodModel(Class<?> originClass, Method method, String name, List<MethodParameterModel> parameters, Type returnType, List<String> comments) {
```
Is 'name' param still needed with the addition of 'method'?

hashtag-smashtag replied:
probably not... but I'm not sure how much I should let the fork diverge, we might want to merge the official master in again at some point

ispringer commented on typescript-generator-core/src/main/java/cz/habarta/typescript/generator/parser/RestMethodModel.java:
```diff
@@ -14,10 +15,10 @@
     private final MethodParameterModel entityParam;
 
-    public RestMethodModel(Class<?> originClass, String name, Type returnType,
+    public RestMethodModel(Class<?> originClass, Method method, String name, Type returnType,
```
Again, maybe 'name' param is not needed?

hashtag-smashtag replied:
same reasoning as above, yeah

ispringer commented on typescript-generator-core/src/test/java/cz/habarta/typescript/generator/CustomSignatureTest.java:
```diff
@@ -0,0 +1,251 @@
+import java.util.stream.Collectors;
+
+@SuppressWarnings("unused")
+public class CustomSignatureTest {
```
Nice tests!

hashtag-smashtag committed:
EVG-9005 can now find include/exclude annotation at declaring class level as well.

 Also added tests and re-organized them so relevant classes/etc are adjacent to the test for better readability.

hashtag-smashtag committed:
fix their flaky test

hashtag-smashtag committed:
EVG-9005 couple more enhancements for our use cases

 * Don't emit non-public methods
 * Support signature annotations on class/interface/enum. Expose parts of CustomSignatureTypeProcessor to Emitter, since current design seems to want DefaultTypeProcessor to return RererenceType in order to actually continue to process the internals (beyond '{') of the class/interface/enum
 * Related tests

hashtag-smashtag committed:
EVG-9005 detect ambiguous signature due to >1 annotation

hashtag-smashtag committed:
EVG-9005 fix SAMTest, method was not public

hashtag-smashtag committed:
EVG-9005 get their tests passing again: avoid NPE and don't check for bean custom signature when origin is null, which happens in some rest app flows
